### PR TITLE
Add CompilerFeatureRequiredAttribute

### PR DESCRIPTION
**Pull request by 333fred**
Fixes Issue https://github.com/dotnet/runtime/issues/66167

### CODE REVIEW

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 1]
```diff
@@ -11275,6 +11275,15 @@ public partial class CompilationRelaxationsAttribute : System.Attribute
...
         public int CompilationRelaxations { get { throw null; } }
     }
+    [System.AttributeUsageAttribute(System.AttributeTargets.All, AllowMultiple=true, Inherited=false)]
+    public sealed partial class CompilerFeatureRequiredAttribute : System.Attribute
+    {
+        public const string RefStructs = "RefStructs";
+        public const string RequiredMembers = "RequiredMembers";
+        public CompilerFeatureRequiredAttribute(string featureName) { }
+        public string FeatureName { get { throw null; } }
+        public bool IsOptional { get { throw null; } set { } }
```
Comment (333fred) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:11285]:
```suggestion
        public bool IsOptional { get { throw null; } init { } }
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/CompilerFeatureRequiredAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        /// </summary>
+        public string FeatureName { get; }
+
+        /// <summary>
+        /// If true, the compiler can choose to allow access to the location where this attribute is applied if it does not understand <see cref="FeatureName"/>.
+        /// </summary>
+        public bool IsOptional { get; init; }
+
+        /// <summary>
+        /// The <see cref="FeatureName"/> used for the ref struct C# feature.
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/CompilerFeatureRequiredAttribute.cs:28]: Nit:
```suggestion
        /// The <see cref="FeatureName"/> used for the ref structs C# feature.
```
?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/CompilerFeatureRequiredAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        /// </summary>
+        public const string RefStructs = nameof(RefStructs);
+
+        /// <summary>
+        /// The <see cref="FeatureName"/> used for the required member C# feature.
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/CompilerFeatureRequiredAttribute.cs:33]: Nit:
```suggestion
        /// The <see cref="FeatureName"/> used for the required members C# feature.
```
?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Runtime/tests/System/Runtime/CompilerServices/AttributesTests.cs [hunk 4]
```diff
@@ -326,5 +326,17 @@ public static void RequiredMemberAttributeTests()
...
+        [Fact]
+        public static void CompilerFeatureRequiredTests()
+        {
+            var attr1 = new CompilerFeatureRequiredAttribute("feature1");
+            Assert.Equal("feature1", attr1.FeatureName);
+            Assert.False(attr1.IsOptional);
+
+            var attr2 = new CompilerFeatureRequiredAttribute("feature2") { IsOptional = true };
+            Assert.Equal("feature2", attr1.FeatureName);
+            Assert.True(attr1.IsOptional);
```
Comment (333fred) [inline src/libraries/System.Runtime/tests/System/Runtime/CompilerServices/AttributesTests.cs:339]:
```suggestion
            Assert.Equal("feature2", attr2.FeatureName);
            Assert.True(attr2.IsOptional);
```

**Pull request merged**